### PR TITLE
Blanks evaluation

### PR DESCRIPTION
**Pull request by maciek-pioro**
# Description
todo

# Checklist
   - [ ] Is this PR focused on a single feature without the possibility of being divided into smaller PRs?
   - [ ] Are all variables set with logical defaults that are backward compatible?
   - [ ] I attached link to neptune if it was necessary
   - [ ] Have you successfully executed the linter and tests?
   - [ ] Are all functions concise and don't require splitting?
   - [ ] Is there documentation for arguments and complex logic?
   - [ ] Are there no temporary solutions in the code? Should a task be added to Trello to refine the code?
   - [ ] Are all functions placed in appropriate files?
   - [ ] Is the PR name meaningful and descriptive?
   - [ ] Is PR description provided, or unnecessary?
  
Other questions:
   - [ ] If I made changes to the requirements or anything related to the Singularity image I did generate new image and upload a new image to the clusters
   - [ ] If this change is significant enough I notified all individuals working with this repository (@channel in slack channel)
   - [ ] If there is a need for a second reviewer I requested additional review 

# Reviewer's checklist:
   - [ ] Does the Neptune link work and does it actually compare the situation before and after the PR?
   - [ ] Do I understand the code?
   - [ ] Do I think another reviewer is not needed or I commented that this PR need another reviewer?

### CODE REVIEW

File: research/conditional/train/configs/baselines/test_env_copy.yaml
Comment (crewtool) [inline research/conditional/train/configs/baselines/test_env_copy.yaml:1]: unrelated / not a baseline / not blanks

**Review by crewtool — DISMISSED** [review]
Nicely done! Please remove vanilla config though

File: research/blanks/test_utils.py [hunk 1]
```diff
@@ -2,6 +2,11 @@
```
Comment (ciebson123) [inline research/blanks/test_utils.py:2]: I really like the fact that tests are included

File: research/blanks/model.py [hunk 2]
```diff
@@ -83,30 +99,54 @@ def get_model(
...
+        return lambda: llm.FeedForward(
+            args.dmodel, args.dff, init_type=args.init_type, init_scale=args.init_scale
+        )
     else:
         raise NotImplementedError(f"ff_mode {args.ff_mode} not implemented")
 
 
-class BlankDiffPredictionHead(nn.Module):
+def get_attention_layer(args):
+    attention_layer_fun = lambda: llm.Attention(
```
Comment (ciebson123) [inline research/blanks/model.py:110]: Note to self: we will need some custom attention here

**Review by ciebson123 — COMMENTED** [review]
Looks great! Does it run?

**Review by Simontwice — COMMENTED** [review]
Could you summarise the design and the choices made?

**Comment by maciek-pioro** [comment]
> Could you summarise the design and the choices made?
@Simontwice  now it's in the description.

File: research/blanks/loss.py [hunk 3]
```diff
@@ -143,11 +141,16 @@ def calculate_llm_loss(
                         assert not input_tokens[nth_blank_mask == 1].eq(blank_id).any()
                     else:
                         assert input_tokens[nth_blank_mask == 1].eq(blank_id).all()
+
                     nth_blank_loss = (nth_blank_mask.reshape(-1) * mask_loss).sum()
+
```
Comment (Simontwice) [inline research/blanks/loss.py:146]: this likely calls for a test for correctness

**Comment by Simontwice** [comment]
It's a sizeable PR, but as long as the following things are tested:

- packing blanks into batch
- finding the last blank
- adjusting the loss mask for the last blank
then it should be perfect. Good job!

File: research/blanks/loss.py (hunk 3 above)
Comment (maciek-pioro) [reply research/blanks/loss.py:146]: Not really the point of this PR, but yeah, we try to have as many tests and asserts as possible in blanx.

**Review by ciebson123 — APPROVED** [review]
:b:est change
